lib/SPIRVProducerPass.cpp

```diff
@@ -1471,19 +1530,60 @@ Type *SPIRVProducerPass::CanonicalType(Type *type) {
   return type;
 }
 
+bool SPIRVProducerPass::PointerRequiresLayout(unsigned aspace) {
+  if (Option::SpvVersion() >= SPIRVVersion::SPIRV_1_4) {
+    switch (aspace) {
+    case AddressSpace::PushConstant:
+    case AddressSpace::Global:
+    case AddressSpace::Constant:
+      return true;
+    default:
+      break;
+    }
+  }
+  return false;
+}
+
 SPIRVID SPIRVProducerPass::getSPIRVType(Type *Ty) {
+  // Prior to 1.4, layout decorations are more relaxed so we can reuse a laid
+  // out type in non-laid out storage classes.
+  bool needs_layout = false;
+  if (auto ptr_ty = dyn_cast<PointerType>(Ty)) {
+    needs_layout = PointerRequiresLayout(ptr_ty->getPointerAddressSpace());
+  }
+  return getSPIRVType(Ty, needs_layout);
+}
+
+SPIRVID SPIRVProducerPass::getSPIRVType(Type *Ty, bool needs_layout) {
+  // Only pointers, structs and arrays should have layout decorations.
+  if (!(isa<PointerType>(Ty) || isa<ArrayType>(Ty) || isa<StructType>(Ty))) {
+    needs_layout = false;
+  }
+  const unsigned layout = needs_layout ? 1 : 0;
```

Comment:
What's the meaning of layout.  

Oh, each type in the TypeMap can have up to two entries. 0 is the one without layout, and 1 is the one with layout.
Suggest a comment here saying something like this.  (I haven't reached the change to the TypeMap decl yet.)